### ISSUE TITLE
[RFC] Python3 Timeline

### ISSUE THREAD

**Issue by tqchen** [dmlc/tvm#1602]
We plan to do some improvements to the IR in this release cycle, and one such question that comes up is that to what extent do we agree on python support. Up until now, TVM has always been compatible to both python2 and python3. However, there are language features in python3 that we like, in particular:

- Type annotations like ```def myfunc(x:int) -> int``` introduced python3.5
- Class type annotations that decorate fields introduced python3.6

Introducing these type annotations to the code will make the program cleaner, but that also means the code will not be python2 compatible. If we introduce class type annotation, we can only use python3.6. One of the major benefits is that adopting type annotation allows the introduction of typed DSLs 

Some of the options are:

- Keep support python2
- For a new set of features, require python3.5+
- Enforce python3.6

For background, currently, most ubuntu version(14.04 and 16.04) comes with python3.5 by default. Most packages plan to stop supporting python2 on 2019

**Comment by tqchen** [comment]
cc @dmlc/tvm-team

**Comment by nhynes** [comment]
+1 for python3.6

* Amazon Linux 2 and [Homebrew (macOS)](https://github.com/Homebrew/homebrew-core/blob/master/Formula/python.rb) are on 3.7
* [Ubuntu 18.04 LTS is on 3.6](https://packages.ubuntu.com/bionic/python/python3)
* the [EPEL](https://mirror.umd.edu/fedora/epel/7/x86_64/Packages/p/) contains python 3.6

**Comment by yzhliu** [comment]
FYI https://python3statement.org/

I tend to have a conservative timeline. Python2 is still installed by default in most Linux distro, supporting py2 is friendly to those just want to have a try. And I'm not sure how many enterprise users are stuck on py2 at the moment.

**Comment by nhynes** [comment]
FWIW, TVM de facto requires LLVM 4+  for CPU support, but LLVM isn't in any of the current distros. Python compiles _much_ faster than LLVM, so I wouldn't take what's in the default repos to be a constraint.

**Comment by tmoreau89** [comment]
+1 for 3.6, it seems like the way forward. We already test script compatibility in integration tests right?

**Comment by masahi** [comment]
Dropping py2 completely seems too aggressive to me. I often see error logs on the discussion forum where people are still using python 2.7. 

So I'm +1 for requiring py3 for new features only for now. In the meantime, we should encourage users to switch to py3, in the next release note for example.

**Comment by tqchen** [comment]
The conclusion seems to be we should always enforce py2 compatibility for the runtime. We might consider requiring py3 for new features

**Comment by FrozenGene** [comment]
+1 For a new set of features, require python3.5+. Enforce python3.6 is too aggressive. Many Linux distro are Python 3.5.

**Comment by zhreshold** [comment]
+1 for 3.5, it's too early to drop 3.5 at this time.

**Comment by junrushao1994** [comment]
I am a big fan of python 3.6+, just a question out of curiosity: typed_ast in mypy seems to support python 2.7 as well, using grammar like
```
some_list = []  # type: List[int]
```
So what prevents us supporting python 2.7?

Thanks!

**Comment by tqchen** [comment]
@junrushao1994 I agree that we could use the comment callback for backward compat with 2.7

**Comment by tqchen** [comment]
see https://github.com/dmlc/tvm/issues/2715
